h264/bits/bitreader.go

```diff
@@ -0,0 +1,119 @@
+  met:
+
+    * Redistributions of source code must retain the above copyright
+  notice, this list of conditions and the following disclaimer.
+    * Redistributions in binary form must reproduce the above
+  copyright notice, this list of conditions and the following disclaimer
+  in the documentation and/or other materials provided with the
+  distribution.
+    * Neither the name of Google Inc. nor the names of its
+  contributors may be used to endorse or promote products derived from
+  this software without specific prior written permission.
+
+  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+  A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+  OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+  SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+  LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+  DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+  THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+  (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+  OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*/
+
+// Package bits provides a bit reader implementation that can read or peek from
+// an io.Reader data source.
+package bits
+
+import (
+	"bufio"
+	"io"
+)
+
+// BitReader is a bit reader that provides methods for reading bits from an
+// io.Reader source.
+type BitReader struct {
+	r    *bufio.Reader
+	n    uint64
+	bits uint
+}
+
+// NewBitReader returns a new BitReader.
+func NewBitReader(r io.Reader) *BitReader {
+	byter, ok := r.(*bufio.Reader)
+	if !ok {
+		byter = bufio.NewReader(r)
+	}
+	return &BitReader{r: byter}
+}
+
+// ReadBits reads n bits from the source and returns as an uint64.
+// For example, with a source as []byte{0x8f,0xe3} (1000 1111, 1110 0011), we
+// would get the following results for consequtive reads with n values:
+// n = 4, res = 0x8 (1000)
+// n = 2, res = 0x3 (0011)
+// n = 4, res = 0xf (1111)
+// n = 6, res = 0x23 (0010 0011)
+func (br *BitReader) ReadBits(bits uint) (uint64, error) {
+	for bits > br.bits {
+		b, err := br.r.ReadByte()
+		if err == io.EOF {
+			return 0, io.ErrUnexpectedEOF
+		}
+		if err != nil {
+			return 0, err
+		}
+		br.n <<= 8
+		br.n |= uint64(b)
+		br.bits += 8
+	}
+
+	n := (br.n >> (br.bits - bits)) & ((1 << bits) - 1)
+	br.bits -= bits
+	return n, nil
+}
+
+// PeekBits provides the next n bits, but without advancing through the source.
+// For example, with a source as []byte{0x8f,0xe3} (1000 1111, 1110 0011), we
+// would get the following results for consequtive peeks with n values:
+// n = 4, res = 0x8 (1000)
+// n = 8, res = 0x8f (1000 1111)
+// n = 16, res = 0x8fe3 (1000 1111, 1110 0011)
+func (br *BitReader) PeekBits(bits uint) (uint64, error) {
```

Comment:
Is there a reason for the `uint` over `int`?

I'd call `bits` here `n`. It reduces stutter and removes the need for the confusing underscore prefix below.

Reply:
shifts need to be uint, so I thought it would avoid conversions.

Reply:
Conversions are free if the size is the same.

"it is true that all things should be made as simple as possible – but no simpler. (And for every problem that is muddled by over-complexity, a dozen are muddled by over-simplifying.)"

— Sydney J. Harris (paraphrasing Einstein, maybe)

Reply:
The bzip2 pkg also use uints, btw. I would have thought it safer to restrict the usage of these to only uints ?

Reply:
Yeah, I know that they do. How many bits can we get? Only 64. We don't have an integer type that is constrained to [0, 64]. So, there's no real win.

Reply:
but if it stops someone using ReadBits or PeekBits with an int that could potentially be -, isn't that a win ? I guess I'm just trying to understand why there are signed and unsigned types if they don't help to avoid a programmer doing something weird.

Reply:
Fixed. Would still like to understand when to use uint and int though. I've always thought that signed and unsigned types helped to find potential problems during compilation, instead of runtime.

Reply:
In this case it doen't make too much of a difference. Knowing when something has underflowed below zero is a valuable thing. That can be known here in both the `int` and `uint` case since there is an upper bound that in smaller than the largest value we accept. The balance for choice then comes down to what is the most ergonomic. `int` wins that.

Interestingly, the bounds check is quicker for uint, since you only have one condition instead of two, but that doesn't matter (if it does - a temporary uint conversion for the check is free).

Reply:
Sure, thanks!